hibernate-core/src/main/java/org/hibernate/query/sqm/sql/BaseSqmToSqlAstConverter.java

```diff
@@ -6562,19 +6562,9 @@ public Object visitBinaryArithmeticExpression(SqmBinaryArithmetic<?> expression)
 		SqmExpression<?> leftOperand = expression.getLeftHandOperand();
 		SqmExpression<?> rightOperand = expression.getRightHandOperand();
 
-		// Need to infer the operand types here first to decide how to transform the expression
-		final FromClauseIndex fromClauseIndex = fromClauseIndexStack.getCurrent();
-		inferrableTypeAccessStack.push( () -> determineValueMapping( rightOperand, fromClauseIndex ) );
-		final MappingModelExpressible<?> leftOperandType = determineValueMapping( leftOperand );
-		inferrableTypeAccessStack.pop();
-		inferrableTypeAccessStack.push( () -> determineValueMapping( leftOperand, fromClauseIndex ) );
-		final MappingModelExpressible<?> rightOperandType = determineValueMapping( rightOperand );
-		inferrableTypeAccessStack.pop();
-
 		final boolean durationToRight = isDuration( rightOperand.getNodeType() );
```

Comment:
I think while playing with this code before, I noticed this line kind of depends on `determineValueMapping` being called on the arguments first. I don't recall exactly what it was that got me into thinking this, but I'm sure it has to do with parameters, since the `bindType` of parameters might change after type determination.

For lack of examples that show a misbehavior, I would say let's just go ahead with your PR and fix issues if something comes up. Just wanted to point out a potential problematic side of this change in case you plan to do further testing.

Reply:
It's been a long time, but from what I remember, if we'd always call `determineValueMapping` for paths which have not yet been processed we would run into this error as the table groups containing them would not yet have been created. That's why, when the temporal type can be determined without inferring the other operand's type, I chose to avoid the call.

Note for myself: since you mentioned this might be problematic with parameters, I might try adding a test which calls `coalesce` on a path and a parameter expression (instead of a literal, like in the test I wrote) and see what happens then.

Reply:
Right so I went back to this issue and tested a bit more, and was able to come up with some cases with parameters where the error I was trying to fix still occurred. I found an alternative solution, and opened https://github.com/hibernate/hibernate-orm/pull/9032 where I implemented that and left the type inference to always be executed like it was before.